File: solutions/java/src/main/java/playground/algorithm/QuickSort.java
package playground.algorithm;

import playground.util.*;

public class QuickSort {

    public void sort(int[] nums) {
        sort(nums, 0, nums.length - 1);
    }

    private void sort(int[] nums, int start, int end) {
        if (start >= end) {
            return;
        }

        int left = start, right = end;
        int key = nums[start];
        while (left < right) {
            while (left < right && nums[right] >= key) {
                right--;
            }
            while (left < right && nums[left] <= key) {
                left++;
            }

            Utils.swap(nums, left, right);
        }
        Utils.swap(nums, start, left);
        sort(nums, start, left - 1);
        sort(nums, left + 1, end);
    }
}

File: solutions/go/src/playground/algorithm/quick_sort.go
package algorithm

func quickSort(array []int) {
	if len(array) <= 1 {
		return
	}

	left, right := 0, len(array)-1
	for left < right {
		for left < right && array[right] >= array[0] {
			right--
		}
		for left < right && array[left] <= array[0] {
			left++
		}
		array[left], array[right] = array[right], array[left]
	}

	array[0], array[left] = array[left], array[0]
	quickSort(array[:left])
	quickSort(array[left+1:])
}
